### PR TITLE
client: unlock client_lock when copying data and do more check for the client_lock

### PR DESCRIPTION
**Pull request by lxbsz**
Unlock client_lock when copying data to bufferlist And avoid using the client_lock.unlock/.lock directly if possible, and add more check for the client_lock.

Fixes: https://tracker.ceph.com/issues/47039
Fixes: https://tracker.ceph.com/issues/47047
Signed-off-by: Xiubo Li <xiubli@redhat.com>
 

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -9610,13 +9609,14 @@ int Client::_read_sync(Fh *f, uint64_t off, uint64_t len, bufferlist *bl,
...
     client_lock.unlock();
     int r = onfinish.wait();
-    client_lock.lock();
 
     // if we get ENOENT from OSD, assume 0 bytes returned
     if (r == -ENOENT)
       r = 0;
-    if (r < 0)
+    if (r < 0) {
+      client_lock.lock();
```
Comment (batrick) [inline src/client/Client.cc:9623]: Can we use a goto instead? This is really awkward and error prone.

File: src/client/Client.cc [hunk 2]
```diff
@@ -9724,18 +9731,20 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
...
+		* does not have enough data to fill in the iov
                 */
                const auto round_size = std::min<unsigned>(resid, iov[j].iov_len);
                iter.copy(round_size, reinterpret_cast<char*>(iov[j].iov_base));
                resid -= round_size;
                /* iter is self-updating */
         }
-        return r;  
+        client_lock.lock();
+        return r;
```
Comment (batrick) [inline src/client/Client.cc:9753]: ditto

File: src/client/Client.h [hunk 3]
```diff
@@ -854,7 +854,7 @@ class Client : public Dispatcher, public md_config_obs_t {
    * Resolve file descriptor, or return NULL.
    */
   Fh *get_filehandle(int fd) {
-    ceph::unordered_map<int, Fh*>::iterator p = fd_map.find(fd);
+    auto p = fd_map.find(fd);
```
Comment (batrick) [inline src/client/Client.h:857]: If you're gonna fix this, please also update the name to `it`. (`p` is short for `std::pair` but this is an iterator!)

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Client.cc (hunk 1 above)
Comment (lxbsz) [reply src/client/Client.cc:9623]: Yeah, sure, will fix it.

File: src/client/Client.cc (hunk 2 above)
Comment (lxbsz) [reply src/client/Client.cc:9753]: sure.

File: src/client/Client.h (hunk 3 above)
Comment (lxbsz) [reply src/client/Client.h:857]: Okay, will fix it.

File: src/client/Client.cc (hunk 2 above)
Comment (lxbsz) [reply src/client/Client.cc:9753]: For this it makes no sense to switch to goto.
We are only doing the unlock/lock in the else branch for read.

File: src/client/Client.cc [hunk 4]
```diff
@@ -335,11 +335,8 @@ Client::~Client()
 void Client::tear_down_cache()
 {
   // fd's
-  for (ceph::unordered_map<int, Fh*>::iterator it = fd_map.begin();
-       it != fd_map.end();
-       ++it) {
-    Fh *fh = it->second;
-    ldout(cct, 1) << __func__ << " forcing close of fh " << it->first << " ino " << fh->inode->ino << dendl;
+  for (auto &[fd, fh] : fd_map) {
+    ldout(cct, 1) << __func__ << " forcing close of fh " << fd << " ino " << fh->inode->ino << dendl;
```
Comment (batrick) [inline src/client/Client.cc:339]: This is a fine change to sneak in but it should be in a separate commit.

File: src/client/Client.cc [hunk 5]
```diff
@@ -5831,6 +5828,8 @@ int Client::authenticate()
 
 int Client::fetch_fsmap(bool user)
 {
+  ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
```
Comment (batrick) [inline src/client/Client.cc:5831]: ditto

File: src/client/Client.cc [hunk 6]
```diff
@@ -8554,9 +8553,9 @@ int Client::readdir_r_cb(dir_result_t *d, add_dirent_cb_t cb, void *p,
 	_ll_get(inode);
       }
 
-      client_lock.unlock();
+      cl.unlock();
       r = cb(p, &de, &stx, next_off, inode);  // _next_ offset
-      client_lock.lock();
+      cl.lock();
```
Comment (batrick) [inline src/client/Client.cc:8558]: readdit changes should also be in a separate commit

File: src/client/Client.h [hunk 7]
```diff
@@ -852,10 +852,10 @@ class Client : public Dispatcher, public md_config_obs_t {
    * Resolve file descriptor, or return NULL.
    */
   Fh *get_filehandle(int fd) {
-    ceph::unordered_map<int, Fh*>::iterator p = fd_map.find(fd);
-    if (p == fd_map.end())
+    auto it = fd_map.find(fd);
+    if (it == fd_map.end())
       return NULL;
-    return p->second;
+    return it->second;
```
Comment (batrick) [inline src/client/Client.h:858]: separate commit for this too

File: src/client/Client.cc [hunk 8]
```diff
@@ -9769,18 +9787,20 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
         if (r <= 0)
           return r;
 
+        client_lock.unlock();
```
Comment (batrick) [inline src/client/Client.cc:9790]: What would be more conventional here would be to pass the `std::unique_lock<>&` to these helper methods so they can manipulate the lock that way.

**Review by batrick — CHANGES_REQUESTED** [review]
Otherwise LGTM

**Comment by lxbsz** [comment]
@batrick 
Fixed them all, thanks.

File: src/client/Client.cc [hunk 9]
```diff
@@ -9769,18 +9785,22 @@ int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
         if (r <= 0)
           return r;
 
+        std::unique_lock cl{client_lock, std::adopt_lock};
```
Comment (batrick) [inline src/client/Client.cc:9788]: This isn't quite what I had in mind. It's weird to adopt the lock because if an exception were to occur, you'd unlock the mutex twice.

Again, the right thing to do here is pass the `std::unique_lock<>&` to the helper method.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/client/Client.cc (hunk 9 above)
Comment (lxbsz) [reply src/client/Client.cc]: Sure, will fix it.

**Comment by batrick** [comment]
````
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:9756:9: error: no declaration matches 'int64_t Client::_preadv_pwritev_locked(Fh*, const iovec*, unsigned int, int64_t, bool, bool, std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >&)'
 9756 | int64_t Client::_preadv_pwritev_locked(Fh *fh, const struct iovec *iov,
      |         ^~~~~~
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:89:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note: candidate is: 'int64_t Client::_preadv_pwritev_locked(Fh*, const iovec*, unsigned int, int64_t, bool, bool)'
 1237 |   int64_t _preadv_pwritev_locked(Fh *f, const struct iovec *iov,
      |           ^~~~~~~~~~~~~~~~~~~~~~
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:89:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:230:7: note: 'class Client' defined here
  230 | class Client : public Dispatcher, public md_config_obs_t {
      |       ^~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc: In member function 'int Client::_preadv_pwritev(int, const iovec*, unsigned int, int64_t, bool)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:9819:75: error: no matching function for call to 'Client::_preadv_pwritev_locked(Fh*&, const iovec*&, unsigned int&, int64_t&, bool&, bool, std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >&)'
 9819 |     return _preadv_pwritev_locked(fh, iov, iovcnt, offset, write, true, cl);
      |                                                                           ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:89:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note: candidate: 'int64_t Client::_preadv_pwritev_locked(Fh*, const iovec*, unsigned int, int64_t, bool, bool)'
 1237 |   int64_t _preadv_pwritev_locked(Fh *f, const struct iovec *iov,
      |           ^~~~~~~~~~~~~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note:   candidate expects 6 arguments, 7 provided
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc: In member function 'int64_t Client::ll_writev(Fh*, const iovec*, int, int64_t)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:13851:70: error: no matching function for call to 'Client::_preadv_pwritev_locked(Fh*&, const iovec*&, int&, int64_t&, bool, bool, std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >&)'
13851 |   return _preadv_pwritev_locked(fh, iov, iovcnt, off, true, false, cl);
      |                                                                      ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:89:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note: candidate: 'int64_t Client::_preadv_pwritev_locked(Fh*, const iovec*, unsigned int, int64_t, bool, bool)'
 1237 |   int64_t _preadv_pwritev_locked(Fh *f, const struct iovec *iov,
      |           ^~~~~~~~~~~~~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note:   candidate expects 6 arguments, 7 provided
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc: In member function 'int64_t Client::ll_readv(Fh*, const iovec*, int, int64_t)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:13861:71: error: no matching function for call to 'Client::_preadv_pwritev_locked(Fh*&, const iovec*&, int&, int64_t&, bool, bool, std::unique_lock<ceph::mutex_debug_detail::mutex_debug_impl<false> >&)'
13861 |   return _preadv_pwritev_locked(fh, iov, iovcnt, off, false, false, cl);
      |                                                                       ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.cc:89:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note: candidate: 'int64_t Client::_preadv_pwritev_locked(Fh*, const iovec*, unsigned int, int64_t, bool, bool)'
 1237 |   int64_t _preadv_pwritev_locked(Fh *f, const struct iovec *iov,
      |           ^~~~~~~~~~~~~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/client/Client.h:1237:11: note:   candidate expects 6 arguments, 7 provided
src/client/CMakeFiles/client.dir/build.make:62: recipe for target 'src/client/CMakeFiles/client.dir/Client.cc.o' failed
make[3]: *** [src/client/CMakeFiles/client.dir/Client.cc.o] Error 1
CMakeFiles/Makefile2:30648: recipe for target 'src/client/CMakeFiles/client.dir/all' failed
````

**Comment by lxbsz** [comment]
Sorry, the `client.h` changes still in the git stash, will update it soon.

**Review by batrick — APPROVED** [review]

**Comment by batrick** [comment]
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200926.000836

failures known/unrelated.
